SourceX/qol.cpp

```diff
@@ -27,6 +42,18 @@ int GetTextWidth(const char *s)
 	return l;
 }
 
+// Fills out with printed number n using ',' thousands separator.
+// Returns pointer to first char after the written number.
+char* PrintWithSeparator(char* out, long long n)
+{
+	if (n < 1000) {
+		return out + sprintf(out, "%lld", n);
+	}
+
+	char* append = PrintWithSeparator(out, n / 1000);
+	return append + sprintf(append, ",%03lld", n % 1000);
+}

```

Comment:
I could come up with arbitrarily fancy ways to do this, but ultimately it just prints a number into a buffer, without allocating memory (unless sprintf is funny).

This is about as simple and straightforward as it gets. There is a POSIX 2008 extension to printf style format strings that allows using the ' character for decimal grouping using locale specific rules, but I really doubt that support for this can be assumed on all platforms.

Reply:
> This is about as simple and straightforward as it gets.

I'm sorry but I don't believe you, even not knowing much about C++ at all. There must be a way to do something like `Print(number, "0.000.000")` or an equivalent thing.

Reply:
https://stackoverflow.com/questions/43482488/how-to-format-a-number-with-thousands-separator-in-c-c

Knock yourself out :D

Reply:
Btw. of all the answers, the most C++ way is probably the locale one.

For someone reading the code, not only does it require that the person has an idea what locales are and how they work, but it also expects to stream the number into a destination - usually something like a stringstream.

If I want to get away without the additional memory allocations this will incur, I have to additionally derive my own memory stream from std::streambuf and use that.

Instead of that, most likely I could throw out the entire char buffer thing, and instead use a std::string, print into that and then fetch the result and throw it into PrintPanelStr, which also does memory allocations along the way...

All of which is massively more convoluted than just printing with that simple function.

Reply:
> I'm sorry but I don't believe you

The way it's done currently is indeed the way to do it from C++11 to C++20 (when not using any support libraries, like `fmt` or Abseil)

There are alternatives but they're bad for other reasons.